CAMEL-12545: create a yaml based route loader

lburgazzoli opened the pull request.

davsclaus commented:
Lets move this into core folder so we dont have build order problem with maven. Its also a core feature as a dsl, like we have for endpoint/component/java/xml already.

lburgazzoli commented:
mh, looks like there is still a cycle as the yaml maven plugin needs the full catalog, maybe we should move it to a dedicated `dsl` module which gets compiled after the catalog ?

davsclaus commented:
ok yeah that sounds okay.

lburgazzoli commented:
@davsclaus PR is now ready

- everything is Java 8 compatible 
- added initial documentation (from camel-k but not yet linked to the DSL section on the website (I may need some help for that))
- cleaned up unused stuffs from camel-k

davsclaus commented:
Actually there are not so much "custom" DSL stuff - there is a bit about saga / doTry etc 

I think all the learnings from this is valuable when / if we go build a universal agnostic route model, and then 100% source code generate the DSL for all supported languages - so they are all compatible and there are no special hacks.

lburgazzoli commented on dsl/camel-yaml-dsl/camel-yaml-dsl-common/src/main/java/org/apache/camel/dsl/yaml/common/YamlSupport.java:
```diff
@@ -0,0 +1,232 @@
+        ObjectHelper.notNull(properties, "properties");
+
+        if (target instanceof CamelContext) {
+            throw new UnsupportedOperationException("Configuring the Camel Context si not supported");
```
done

lburgazzoli commented on dsl/camel-yaml-dsl/camel-yaml-dsl-deserializers/src/generated/java/org/apache/camel/dsl/yaml/deserializers/EndpointConsumerDeserializersResolver.java:
```diff
@@ -0,0 +1,244 @@
+import org.apache.camel.dsl.yaml.common.YamlDeserializerResolver;
+import org.snakeyaml.engine.v2.api.ConstructNode;
+
+public final class EndpointConsumerDeserializersResolver implements YamlDeserializerResolver {
```
added as file comment

lburgazzoli commented:
> I think all the learnings from this is valuable when / if we go build a universal agnostic route model, and then 100% source code generate the DSL for all supported languages - so they are all compatible and there are no special hacks.

Absolutely, I also have some plans to see if we can make the definitions for which I had to hand write some decoders more DSL friendly so we can probably get rid of them.

lburgazzoli commented:
ok I have removed some of the custom deserializers except those needed for input or error handling.

The PR it is now ready for final review. 

I had to introduce  a new annotation '@DslProperty' which may need a better name to properly identify some fields marked with '@XmlTransient' that should not be ignored by alternative DSLs.
